`panic: assignment to entry in nil map`, when exposing a Skupper service used as a statefulset base service, in headless mode.

Steps to reproduce:

-    Create a skupper service named 'backend'
-    Create a statefulset, whose spec has `ServiceName` as `backend`
-    Expose the statefulset as headless

While this is not the normal way of exposing a statefulset (ie, there was first a `service create` then an `expose statefulset`, instead of a `service bind`), the panic is unexpected.

Also, note that the issue is not related to the new '--publish-not-ready-addresses' flag, as it happens with or without it, whenever `–headless` is used.


```
$ skupper --namespace private-issue-832-1 expose statefulset backend --headless --publish-not-ready-addresses
panic: assignment to entry in nil map
goroutine 1 [running]:
github.com/skupperproject/skupper/client.(*VanClient).GetHeadlessServiceConfiguration(0xc00041f3b0, {0x7fff7dce1fee, 0x7}, {0x55eef7ed16cc, 0x3}, {0x0, 0x0}, {0x55eef8eb1bf0, 0x0, 0x0}, ...)
    /builddir/build/BUILD/skupper-1.2.0/client/serviceinterface_update.go:238 +0xb45
main.expose({0x55eef8585f20, _}, {_, _}, {_, _}, {_, _}, {{0x55eef7ed16cc, 0x3}, ...})
    /builddir/build/BUILD/skupper-1.2.0/cmd/skupper/skupper.go:204 +0x553
main.(*SkupperKubeService).Expose(0xc000010c20, 0xc0004d9db0?, {0xc0004802a0?, 0x0?, 0x0?})
    /builddir/build/BUILD/skupper-1.2.0/cmd/skupper/skupper_kube_expose.go:70 +0x294
github.com/spf13/cobra.(*Command).execute(0xc000476000, {0xc000480180, 0x6, 0x6})
    /builddir/build/BUILD/skupper-1.2.0/vendor/github.com/spf13/cobra/command.go:840 +0x67c
github.com/spf13/cobra.(*Command).ExecuteC(0xc0003ab080)
    /builddir/build/BUILD/skupper-1.2.0/vendor/github.com/spf13/cobra/command.go:945 +0x37c
github.com/spf13/cobra.(*Command).Execute(...)
    /builddir/build/BUILD/skupper-1.2.0/vendor/github.com/spf13/cobra/command.go:885
main.main()
    /builddir/build/BUILD/skupper-1.2.0/cmd/skupper/skupper.go:991 +0x25

$ skupper --namespace private-issue-832-1 expose statefulset backend --headless 
panic: assignment to entry in nil map
goroutine 1 [running]:
github.com/skupperproject/skupper/client.(*VanClient).GetHeadlessServiceConfiguration(0xc000439400, {0x7fff9d45700c, 0x7}, {0x5595bb9a86cc, 0x3}, {0x0, 0x0}, {0x5595bc988bf0, 0x0, 0x0}, ...)
    /builddir/build/BUILD/skupper-1.2.0/client/serviceinterface_update.go:238 +0xb45
main.expose({0x5595bc05cf20, _}, {_, _}, {_, _}, {_, _}, {{0x5595bb9a86cc, 0x3}, ...})
    /builddir/build/BUILD/skupper-1.2.0/cmd/skupper/skupper.go:204 +0x553
main.(*SkupperKubeService).Expose(0xc00012cc10, 0xc00051fdb0?, {0xc0004393b0?, 0x0?, 0x0?})
    /builddir/build/BUILD/skupper-1.2.0/cmd/skupper/skupper_kube_expose.go:70 +0x294
github.com/spf13/cobra.(*Command).execute(0xc00048a000, {0xc000439360, 0x5, 0x5})
    /builddir/build/BUILD/skupper-1.2.0/vendor/github.com/spf13/cobra/command.go:840 +0x67c
github.com/spf13/cobra.(*Command).ExecuteC(0xc0003bd080)
    /builddir/build/BUILD/skupper-1.2.0/vendor/github.com/spf13/cobra/command.go:945 +0x37c
github.com/spf13/cobra.(*Command).Execute(...)
    /builddir/build/BUILD/skupper-1.2.0/vendor/github.com/spf13/cobra/command.go:885
main.main()
    /builddir/build/BUILD/skupper-1.2.0/cmd/skupper/skupper.go:991 +0x25

$ skupper --namespace private-issue-832-1 expose statefulset backend 
statefulset backend exposed as backend
$ 


```

--- a/client/serviceinterface_update.go
+++ b/client/serviceinterface_update.go
@@ -248,6 +248,13 @@ func (cli *VanClient) ServiceInterfaceBind(ctx context.Context, service *types.S
 
 func (cli *VanClient) GetHeadlessServiceConfiguration(targetName string, protocol string, address string, ports []int, publishNotReadyAddresses bool, namespace string) (*types.ServiceInterface, error) {
 	svcNamespace := utils.GetOrDefault(namespace, cli.GetNamespace())
+
+	proxySvcName := targetName + "-proxy"
+	proxyService, err := cli.KubeClient.CoreV1().Services(svcNamespace).Get(context.TODO(), proxySvcName, metav1.GetOptions{})
+	if err == nil && proxyService != nil {
+		return nil, fmt.Errorf("service %s already exists in the cluster, use any other name to expose the statefulset with a headless service", proxySvcName)
+	}
+
 	statefulset, err := cli.KubeClient.AppsV1().StatefulSets(svcNamespace).Get(context.TODO(), targetName, metav1.GetOptions{})
 	if err == nil {
 		if address != "" && address != statefulset.Spec.ServiceName {
@@ -278,6 +285,9 @@ func (cli *VanClient) GetHeadlessServiceConfiguration(targetName string, protoco
 						def.Ports = append(def.Ports, int(port.Port))
 						if port.TargetPort.IntValue() != 0 && int(port.Port) != port.TargetPort.IntValue() {
 							// TODO: handle string ports
+							if def.Headless.TargetPorts == nil {
+								def.Headless.TargetPorts = make(map[int]int)
+							}
 							def.Headless.TargetPorts[int(port.Port)] = port.TargetPort.IntValue()
 						}
 					}

--- a/cmd/service-controller/controller.go
+++ b/cmd/service-controller/controller.go
@@ -95,14 +95,6 @@ func hasProxyAnnotation(service corev1.Service) bool {
 	}
 }
 
-func getProxyName(name string) string {
-	return name + "-proxy"
-}
-
-func getServiceName(name string) string {
-	return strings.TrimSuffix(name, "-proxy")
-}
-
 func hasSkupperAnnotation(service corev1.Service, annotation string) bool {
 	_, ok := service.ObjectMeta.Annotations[annotation]
 	return ok

--- a/cmd/service-controller/controller.go
+++ b/cmd/service-controller/controller.go
@@ -674,6 +674,9 @@ func (c *Controller) ensureHeadlessProxyFor(bindings *service.ServiceBindings, s
 	}
 
 	_, err = kube.CheckProxyStatefulSet(images.GetRouterImageDetails(), serviceInterface, statefulset, config, c.vanClient.Namespace, c.vanClient.KubeClient)
+	if err != nil {
+		event.Recordf(ServiceControllerError, "Error creating new proxy stateful set: %s", err)
+	}
 	return err
 }
 
@@ -685,6 +688,9 @@ func (c *Controller) createHeadlessProxyFor(bindings *service.ServiceBindings) e
 	}
 
 	_, err = kube.NewProxyStatefulSet(images.GetRouterImageDetails(), serviceInterface, config, c.vanClient.Namespace, c.vanClient.KubeClient)
+	if err != nil {
+		event.Recordf(ServiceControllerError, "Error creating new proxy stateful set: %s", err)
+	}
 	return err
 }
 

--- a/pkg/kube/deployments.go
+++ b/pkg/kube/deployments.go
@@ -116,6 +116,11 @@ func NewProxyStatefulSet(image types.ImageDetails, serviceInterface types.Servic
 	svcName := serviceInterface.Address
 	if serviceInterface.IsOfLocalOrigin() {
 		svcName += "-proxy"
+		proxyService, err := cli.CoreV1().Services(namespace).Get(context.TODO(), svcName, metav1.GetOptions{})
+		if err == nil && proxyService != nil {
+			return nil, fmt.Errorf("service %s already exists in the cluster, use any other name to expose the statefulset with a headless service", svcName)
+		}
+
 		_, err = NewHeadlessService(svcName, serviceInterface.Address, serviceInterface.Ports, serviceInterface.Ports, serviceInterface.Labels, &ownerRef, namespace, cli)
 		if err != nil {
 			return nil, err

--- a/pkg/kube/services.go
+++ b/pkg/kube/services.go
@@ -47,16 +47,6 @@ func GetService(name string, namespace string, kubeclient kubernetes.Interface) 
 	return current, err
 }
 
-func NewHeadlessServiceForAddress(address string, ports []int, targetPorts []int, labels map[string]string, owner *metav1.OwnerReference, namespace string, kubeclient kubernetes.Interface) (*corev1.Service, error) {
-	selector := map[string]string{
-		"internal.skupper.io/service": address,
-	}
-	service := makeServiceObjectForAddress(address, ports, targetPorts, labels, selector, owner)
-	service.Spec.ClusterIP = "None"
-
-	return createServiceFromObject(service, namespace, kubeclient)
-}
-
 func NewHeadlessService(name string, address string, ports []int, targetPorts []int, labels map[string]string, owner *metav1.OwnerReference, namespace string, kubeclient kubernetes.Interface) (*corev1.Service, error) {
 	selector := map[string]string{
 		"internal.skupper.io/service": address,

